### PR TITLE
add adjuster property to always apply timestamp and duration if missing

### PR DESCRIPTION
**Pull request by naoman**
PR for #25 
Adding a property to Finagle adjuster to allow applying timestamp and duration without looking at the binary annotations. This rectifies the [finagle Span model issue](https://github.com/openzipkin/zipkin-finagle/issues/10) for non-finagle libraries that use finagle model for instrumentation.

### CODE REVIEW

File: adjuster/finagle/src/main/java/zipkin/sparkstreaming/adjuster/finagle/FinagleAdjuster.java [hunk 1]
```diff
@@ -44,21 +47,30 @@ public static Builder newBuilder() {
      */
     Builder applyTimestampAndDuration(boolean applyTimestampAndDuration);
 
+    /**
+     * This back fills timestamp and duration for Spans coming from Finagle Span model, even if
+     * they are generated by non-finagle libraries.
+     */
+    Builder alwaysApplyTimestampAndDuration(boolean alwaysApplyTimestampAndDuration);
```
Comment (naoman) [inline adjuster/finagle/src/main/java/zipkin/sparkstreaming/adjuster/finagle/FinagleAdjuster.java]: Property name is not very intuitive. Any suggestions?

**Comment by naoman** [comment]
cc @adriancole

**Comment by adriancole** [comment]
This sounds like a different adjuster to me since it hasnt to do with
finagle. Thats probably why it sounds unintuitive.

You could put this in a new adjuster class the default module since it is
generic, maybe note the issue on openzipkin.github.io about timestamp and
duration for context since we usually shouldnt be doing this.

For naming, try to follow the same name for this adjuster as the timestamp
property in finagle so that when you read down the config it looks sensible.

Sound good?

**Comment by naoman** [comment]
I think we should keep this in the same adjuster class (FinagleAdjuster) for following two reasons:
1. We are making an adjustment for [Finagle Span model](https://github.com/openzipkin/zipkin-finagle/issues/10) shortcomings, so anyone who was using Finagle Span model for library instrumentation will need this. 
2. The adjustment is exactly same as the "apply-timestamp-and-duration" adjustment, just the condition to apply is different. So its better to reuse the code, instead of having multiple copies.

**Comment by adriancole** [comment]
You are assuming that the lack of timestamp and duration is finagle centric
when we know with certainty it is not. We cataloged many independent
projects which are not finagle and there is no established "finagle model"

Adding a hook like "timestamp but not for finagle" to a finagle project
sounds like spahagetti to me.


On 15 Mar 2017 08:23, "Naoman Abbas" <notifications@github.com> wrote:

I think we should keep this in the same adjuster class (FinagleAdjuster)
for following two reasons:

   1. We are making an adjustment for Finagle Span model
   <https://github.com/openzipkin/zipkin-finagle/issues/10> shortcomings,
   so anyone who was using Finagle Span model for library instrumentation will
   need this.
   2. The adjustment is exactly same as the "apply-timestamp-and-duration"
   adjustment, just the condition to apply is different. So its better to
   reuse the code, instead of having multiple copies.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openzipkin/zipkin-sparkstreaming/pull/36#issuecomment-286651963>,
or mute the thread

**Comment by naoman** [comment]
While lack of timestamp and duration is not finagle centric, it is an issue with finagle. Even before this change, we were adjusting timestamp and duration for spans coming form finagle libraries. With this change, all we are doing is extending it to the spans that are created based on finagle span model. Finagle is the common factor here. 

I understand that other libraries can also have this issue, but lets tackle that when we run into them. Maybe they need a different handling than this. In case of finagle, we know that the "adjustment" is exactly same in both cases (spans coming from finagle library, and spans coming based on finagle model).

I've updated the property name to make it explicit.

**Comment by adriancole** [comment]
I don't agree at all with this change. finagleSpanModel is jargon I've not heard of before. 

https://github.com/openzipkin/openzipkin.github.io/issues/49 has details on many libraries which all need to accomodate timestamp and duration. This is not a well investigated topic, it isn't at all hypothetical.

On the other hand the "non-finagle" libraries you are hinting at must not be open source, which hints to me that using a custom adjuster is probably the way out. Finagle always adds finagle version properties and the point of this module is to address the finagle library.

**Comment by adriancole** [comment]
here's a summary of the approach I mentioned for how to address libraries that simply don't report timestamp or duration (regardless of how finagly) https://github.com/openzipkin/zipkin-sparkstreaming/issues/37

**Pull request closed without merging**